drivers/SmartThings/zigbee-window-treatment/src/mic/init.lua

```diff
@@ -0,0 +1,187 @@
+    end
+end
+
+local set_window_shade_level = function(level)
+
+    return function(driver, device, cmd)
+        device:send_to_component(cmd.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+    end
+end
+
+local WindowCovering = zcl_clusters.WindowCovering
+
+local SHADE_SET_STATUS = "shade_set_status"
+
+local function current_position_attr_handler(driver, device, value, zb_rx)
+
+    local level = value.value
+    local current_level = device:get_latest_state("main", capabilities.windowShadeLevel.ID,
+        capabilities.windowShadeLevel.shadeLevel.NAME)
+    local windowShade = capabilities.windowShade.windowShade
+
+    if level == 0 then
+        device:emit_event(windowShade.closed())
+        device:emit_event(capabilities.windowShadeLevel.shadeLevel(0))
+    elseif level == 100 then
+        device:emit_event(windowShade.open())
+        device:emit_event(capabilities.windowShadeLevel.shadeLevel(100))
+    else
+        if current_level ~= level or current_level == nil then
+            current_level = current_level or 0
+            device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+            local event = nil
+
+            if current_level ~= level then
+                --  event = current_level < level and windowShade.opening() or windowShade.closing()
+                event = current_level < level and windowShade.partially_open() or windowShade.partially_open()
+            end
+            if event ~= nil then
+                device:emit_event(event)
+            end
+        end
+
+        local set_status_timer = device:get_field(SHADE_SET_STATUS)
+        if set_status_timer then
+            device.thread:cancel_timer(set_status_timer)
+            device:set_field(SHADE_SET_STATUS, nil)
+        end
+        local set_window_shade_status = function()
+            local current_level = device:get_latest_state("main", capabilities.windowShadeLevel.ID,
+                capabilities.windowShadeLevel.shadeLevel.NAME)
+
+            if current_level == 0 then
+                device:emit_event(windowShade.closed())
+            elseif current_level == 100 then
+                device:emit_event(windowShade.open())
+            else
+                device:emit_event(windowShade.partially_open())
+            end
+        end
+
+        set_status_timer = device.thread:call_with_delay(1, set_window_shade_status)
+        device:set_field(SHADE_SET_STATUS, set_status_timer)
+    end
+end
+
+local function window_shade_level_cmd(driver, device, command)
+
+    local level = utils.clamp_value(command.args.shadeLevel, 0, 100)
+    device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+end
+
+local function set_shade_level(device, value, command)
+
+    local level = value
+    device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+end
```

Comment:
these two functions are basically identical I would refactor to remove one of them